controllers/imagerepository_controller.go

```diff
@@ -230,6 +234,139 @@ func (r *ImageRepositoryReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	return ctrl.Result{}, nil
 }
 
+func (r *ImageRepositoryReconciler) NotificationExistsInQuayByUUID(UUID string, imageRepository *imagerepositoryv1alpha1.ImageRepository) (quay.Notification, error) {
+	notification := quay.Notification{}
+	allNotifications, err := r.QuayClient.GetNotifications(r.QuayOrganization, imageRepository.Spec.Image.Name)
+	if err != nil {
+		return notification, err
+	}
+	for _, quayNotification := range allNotifications {
+		if quayNotification.UUID == UUID {
+			notification = quayNotification
+			break
+		}
+	}
+
+	return notification, nil
+}
+
+func (r *ImageRepositoryReconciler) NotificationExistsInQuayByTitle(Title string, imageRepository *imagerepositoryv1alpha1.ImageRepository) (quay.Notification, error) {
+	notification := quay.Notification{}
+	allNotifications, err := r.QuayClient.GetNotifications(r.QuayOrganization, imageRepository.Spec.Image.Name)
+	if err != nil {
+		return notification, err
+	}
+	for _, quayNotification := range allNotifications {
+		if quayNotification.Title == Title {
+			notification = quayNotification
+			break
+		}
+	}
+
+	return notification, nil
+}
+
+func (r *ImageRepositoryReconciler) HandleNotifications(ctx context.Context, imageRepository *imagerepositoryv1alpha1.ImageRepository) error {
+	log := ctrllog.FromContext(ctx).WithName("HandleNotifications")
+
+	if imageRepository.Status.Notifications == nil {
+		// No status notifications to check
+		return nil
+	}
+	log.Info("Handling notifications")
+	for index := 0; index < len(imageRepository.Status.Notifications); index++ {
+		statusNotification := &imageRepository.Status.Notifications[index] // This way we can update the UUID if notification gets updated
+		existsInSpec := false
+		for _, notification := range imageRepository.Spec.Notifications {
+			if notification.Title == statusNotification.Title {
+				existsInSpec = true
+				quayNotification, err := r.NotificationExistsInQuayByUUID(statusNotification.UUID, imageRepository)
```

Comment:
I get your point. Let me think on an alternative here.

Reply:
Note that call to _NotificationExistsInQuayByUUID_ or _NotificationExistsInQuayByTitle_ is performed only once per iteration (= per notification). It depends on the if else where they are placed. It couldn't happen that we call both in the same iteration.

Besides that, you are 100% right in the fact that we should only call Quay's _getNotifications_ only once per _HandleNotifications_ function call. I'll refactor it to avoid call it several times and let _NotificationExistsInQuayByUUID_ and _NotificationExistsInQuayByTitle_ only do comparisons based on the attribute and not calling _getNotifications_ within them.